File: Python3/2865_beautiful_tower_i.py
# https://leetcode.cn/problems/beautiful-towers-i/

class Solution:
    def maximumSumOfHeights(self, maxHeights):
        """
        Inputs:
            maxHeights: list[int]
        Outputs:
            res: int
        """
        n = len(maxHeights)
        res = 0
        for i in range(n):
            cur = maxHeights[i]
            tmp = cur
            # Left.
            for j in range(i-1, -1, -1):
                cur = min(maxHeights[j], cur)
                tmp += cur
            # Right.
            cur = maxHeights[i]
            for j in range(i+1, n):
                cur = min(maxHeights[j], cur)
                tmp += cur
            res = max(res, tmp)
        return res
    
sol = Solution()
maxHeights = [6,5,3,9,2,7] # 22
# maxHeights = [3,2,5,5,2,3] # 18
print(sol.maximumSumOfHeights(maxHeights))

File: Cpp/2865_beautiful_tower_i.cpp
// https://leetcode.cn/problems/beautiful-towers-i/

class Solution {
public:
    long long maximumSumOfHeights(vector<int>& maxHeights) {
        int n = maxHeights.size();
        long long res = 0;
        for (int i = 0; i < n; ++i) {
            int cur = maxHeights[i];
            long long tmp = cur;
            // Left.
            for (int j = i-1; j >= 0; --j) {
                cur = min(cur, maxHeights[j]);
                tmp += cur;
            }
            // Right.
            cur = maxHeights[i];
            for (int j = i+1; j < n; ++j) {
                cur = min(cur, maxHeights[j]);
                tmp += cur;
            }
            res = max(res, tmp);
        }
        return res;
    }
};
